### PR TITLE
Don't update api-reference docs if the only changes are the timestamps

### LINKED ISSUE #16121: Do not update api-reference docs if last updated timestamp is the only change
Right now, rerunning, update-api-reference-docs, keeps updating the docs with new last updated timestamp.
Similar to other generated docs, we should detect that case and not update the doc.

Note that verify-api-reference-docs works fine. (It ignores last update line change)

cc @caesarxuchao

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #16121

cc @jlowdermilk because I changed the REPO_DIR.
cc @lavalamp

### CODE REVIEW

**Comment by jlowdermilk** [comment]
Looks fine, script still reuses existing REPO_DIR. Jenkins unit/integration needs to pass in REPO_DIR to be able to run `hack/verify-api-reference-docs.sh`, so as long as test passes, lgtm.

File: hack/update-api-reference-docs.sh [hunk 1]
```diff
@@ -37,12 +47,40 @@ echo "Reading swagger spec from: ${SWAGGER_PATH}"
...
+    generated=$(cat "${OUTPUT_TEMP}/${file}")
+
+    # Filter out meaningless lines with timestamps
+    original=$(echo "${original}" | grep -v "Last updated" || :)
+    generated=$(echo "${generated}" | grep -v "Last updated" || :)
+
+    # By now, the contents should be normalized and stripped of any
+    # auto-managed content.  
+    if diff -Bw >/dev/null <(echo "${original}") <(echo "${generated}"); then
+      # actual contents same, overwrite generated with original.
```
Comment (nikhiljindal) [inline hack/update-api-reference-docs.sh]: We should overwrite when the contents are different right?

Comment (nikhiljindal) [inline hack/update-api-reference-docs.sh]: aah ok. I see now.
Ignore

**Comment by nikhiljindal** [comment]
LGTM, thanks!

File: hack/update-api-reference-docs.sh (hunk 1 above)
Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: Yes, that's confusing. I copied this pattern from another hack/*.sh.

**Comment by caesarxuchao** [comment]
Restart the checks as I just upload the v4 image to gcr.

**Comment by jlowdermilk** [comment]
Looks like this breaks dockerized test. You can repro locally by running
```
WORKSPACE=/tmp hack/jenkins/gotest-dockerized.sh
```

File: hack/update-api-reference-docs.sh [hunk 2]
```diff
@@ -37,12 +47,40 @@ echo "Reading swagger spec from: ${SWAGGER_PATH}"
 mkdir -p $V1_PATH
 mkdir -p $V1BETA1_PATH
 
-docker run --rm -v $V1_PATH:/output:z -v ${SWAGGER_PATH}:/swagger-source:z gcr.io/google_containers/gen-swagger-docs:v3 \
+docker run -u $(id -u) --rm -v $V1_TEMP_PATH:/output:z -v ${SWAGGER_PATH}:/swagger-source:z gcr.io/google_containers/gen-swagger-docs:v4 \
```
Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: The output is supposed to be written to $V1_TEMP_PATH, but when running in dockerized test, nothing is written.

Comment (jlowdermilk) [inline hack/update-api-reference-docs.sh]: Ah, well there's your problem. The path that you pass to `docker run -v PATH:/output` will get resolved on the host, not the container. So it's actually writing to `${KUBE_TEMP}` on your machine. See longer explanation [here](https://github.com/kubernetes/kubernetes/pull/16287#discussion_r43333840).

The tl;dr is that when you call `docker run -v PATH:/containerpath ...`, the `PATH` must be resolvable on the host. To have the above command write output that you can then read here, you have to pass in the host-resolvable path (i.e., use REPO_DIR), and then read that output using the local path (*without* REPO_DIR).

Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: Thanks @jlowdermilk! So I didn't understand your comments when I reviewed your PR, now it all makes sense!

As you said, the address I passed to -v will be resolved as a host-resolvable path, then we still need to use ${KUBE_TEMP}, there can be race condition if two dockerized tests are running at the same time. Am I right?

Comment (jlowdermilk) [inline hack/update-api-reference-docs.sh]: Test runs don't share workspaces (the jenkins dir where the repo gets cloned down, from which dockerized tests run, mapping in REPO_DIR), so there's no race.

Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: Great. That makes things easier.

File: hack/update-api-reference-docs.sh [hunk 3]
```diff
@@ -22,27 +22,74 @@ set -o nounset
...
-echo "Generating api reference docs at ${OUTPUT}"
+DEFAULT_OUTPUT="${KUBE_ROOT}/docs/api-reference"
+OUTPUT=${1:-${DEFAULT_OUTPUT}}
+# Use REPO_DIR if provided so we can set it to the host-resolvable path
+# to the repo root if we are running this script from a container with
+# docker mounted in as a volume.
+# We pass the host output dir as the source dir to `docker run -v`, but use
+# the regular one to compute diff (they will be the same if running this
+# test on the host, potentially different if running in a container).
+REPO_DIR=${REPO_DIR:-"${KUBE_ROOT}"}
```
Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: @jlowdermilk I moved the reference to REPO_DIR from the verify script to the update script.

File: hack/update-api-reference-docs.sh [hunk 4]
```diff
@@ -22,27 +22,74 @@ set -o nounset
...
+TMP_SUBPATH="_output/generated_html"
+OUTPUT_TMP_IN_HOST="${REPO_DIR}/${TMP_SUBPATH}"
```
Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: Variable has suffix _IN_HOST are subpath of ${REPO_DIR}.

**Comment by caesarxuchao** [comment]
@jlowdermilk, updated. It passed my local dockerized tests. PTAL. Thank you very much for your help!

File: hack/verify-api-reference-docs.sh [hunk 5]
```diff
@@ -26,15 +26,8 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
...
-# to the repo root if we are running this script from a container with
-# docker mounted in as a volume.
-# We pass the host output dir to update-api-reference-docs.sh, but use
-# the regular one to compute diff (they will be the same if running this
-# test on the host, potentially different if running in a container).
-REPO_DIR=${REPO_DIR:-"${KUBE_ROOT}"}
-HOST_OUTPUT_DIR="${REPO_DIR}/_tmp/api-reference"
-TMP_OUTPUT_DIR="${KUBE_ROOT}/_tmp/api-reference"
+HOST_OUTPUT_DIR="${KUBE_ROOT}/_tmp/api-reference"
+mkdir -p ${HOST_OUTPUT_DIR}
```
Comment (jlowdermilk) [inline hack/verify-api-reference-docs.sh]: I would recommend calling this something else, since it's not a host resolvable path. OUTPUT_DIR is fine.

Comment (caesarxuchao) [inline hack/verify-api-reference-docs.sh]: Sure.

**Comment by jlowdermilk** [comment]
lgtm

**Pull request merged**